### ISSUE TITLE
loss

### ISSUE THREAD

**Issue by jl749** [jl749/pytorch-YOLOv4#1]
https://github.com/jl749/pytorch-YOLOv4/blob/a65d219f9066bae4e12003bd7cdc04531860c672/train.py#L134-L165
```
[(w / strides[0], h / strides[0]) for w, h in anchors] 
--> [(1.5, 2.0), (2.375, 4.5), (5.0, 3.5), (4.5, 9.375), (9.5, 6.875), (9.0, 18.25), (17.75, 13.75), (24.0, 30.375), (57.375, 50.125)]
```

```
np.array([all_anchors_grid[j] for j in anch_masks[0]], dtype=np.float32)
--> array([[1.5  , 2.   ],
           [2.375, 4.5  ],
           [5.   , 3.5  ]], dtype=float32)
```

```
np.array(all_anchors_grid, dtype=np.float32)
--> array([[ 1.5  ,  2.   ],
           [ 2.375,  4.5  ],
           [ 5.   ,  3.5  ],
           [ 4.5  ,  9.375],
           [ 9.5  ,  6.875],
           [ 9.   , 18.25 ],
           [17.75 , 13.75 ],
           [24.   , 30.375],
           [57.375, 50.125]], dtype=float32)
```

```
ref_anchors[:, 2:] = np.array(all_anchors_grid, dtype=np.float32)
--> array([[ 0.   ,  0.   ,  1.5  ,  2.   ],
       [ 0.   ,  0.   ,  2.375,  4.5  ],
       [ 0.   ,  0.   ,  5.   ,  3.5  ],
       [ 0.   ,  0.   ,  4.5  ,  9.375],
       [ 0.   ,  0.   ,  9.5  ,  6.875],
       [ 0.   ,  0.   ,  9.   , 18.25 ],
       [ 0.   ,  0.   , 17.75 , 13.75 ],
       [ 0.   ,  0.   , 24.   , 30.375],
       [ 0.   ,  0.   , 57.375, 50.125]], dtype=float32)
```

```
fsize = image_size // strides[0]
--> 608//8 = 76
```

## represents each cell (BATCH_SIZE, ANCHOR_SIZE(usually 3), S_i, S_i)
```
grid_x = torch.arange(fsize, dtype=torch.float).repeat(BATCH_SIZE, 3, fsize, 1)
--> (BATCH_SIZE, 3, 76, 76)
tensor([[[[ 0.,  1.,  2.,  ..., 73., 74., 75.],
          [ 0.,  1.,  2.,  ..., 73., 74., 75.],
          [ 0.,  1.,  2.,  ..., 73., 74., 75.],
          ...,

grid_y = torch.arange(fsize, dtype=torch.float).repeat(batch, 3, fsize, 1).permute(0, 1, 3, 2)
--> tensor([[[[ 0.,  0.,  0.,  ...,  0.,  0.,  0.],
          [ 1.,  1.,  1.,  ...,  1.,  1.,  1.],
          [ 2.,  2.,  2.,  ...,  2.,  2.,  2.],
          ...,
```

```
anchor_w = torch.from_numpy(masked_anchors[:, 0]).repeat(batch, fsize, fsize, 1).permute(0, 3, 1, 2)
anchor_h = torch.from_numpy(masked_anchors[:, 1]).repeat(batch, fsize, fsize, 1).permute(0, 3, 1, 2)

--> (2, 3, 76, 76)
```
<details>
<summary>values</summary>

```
anchor_w = tensor([[[[1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          ...,
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000]],

         [[2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          ...,
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750]],

         [[5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          ...,
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000]]],


        [[[1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          ...,
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000],
          [1.5000, 1.5000, 1.5000,  ..., 1.5000, 1.5000, 1.5000]],

         [[2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          ...,
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750],
          [2.3750, 2.3750, 2.3750,  ..., 2.3750, 2.3750, 2.3750]],

         [[5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          ...,
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000],
          [5.0000, 5.0000, 5.0000,  ..., 5.0000, 5.0000, 5.0000]]]])
anchor_h = tensor([[[[2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          ...,
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000]],

         [[4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          ...,
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000]],

         [[3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          ...,
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000]]],


        [[[2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          ...,
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000],
          [2.0000, 2.0000, 2.0000,  ..., 2.0000, 2.0000, 2.0000]],

         [[4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          ...,
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000],
          [4.5000, 4.5000, 4.5000,  ..., 4.5000, 4.5000, 4.5000]],

         [[3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          ...,
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000],
          [3.5000, 3.5000, 3.5000,  ..., 3.5000, 3.5000, 3.5000]]]])
```

</details>
```
self.masked_anchors.append(masked_anchors)
self.ref_anchors.append(ref_anchors)
self.grid_x.append(grid_x)
self.grid_y.append(grid_y)
self.anchor_w.append(anchor_w)
self.anchor_h.append(anchor_h)
```

**Comment by jl749** [comment]
# forward
https://github.com/jl749/pytorch-YOLOv4/blob/a65d219f9066bae4e12003bd7cdc04531860c672/train.py#L235-L274
![image](https://user-images.githubusercontent.com/67103130/152285621-56bb8e67-53b2-4449-8c6e-66e27108bee8.png)

https://github.com/jl749/pytorch-YOLOv4/blob/a65d219f9066bae4e12003bd7cdc04531860c672/train.py#L248-L252
![image](https://user-images.githubusercontent.com/67103130/150052753-e9df8a75-ed55-4cf1-a7fd-fc9d8d8ede50.png)

**Comment by jl749** [comment]
# build_target
build_target is the most important part of the error calculation, its function is matching, that is, matching the ground-truth box and the predicted box responsible for predicting the ground-truth box. Its main process is as follows:

a. First convert the normalized ground-truth box to the ground-truth box on the feature map size.
b. Calculate the iou of the predefined anchor and the real box, obtain the maximum iou and its index of each real box, and then get the anchor of which size each real box needs to use to match.
c. According to the grid position of the real frame and the anchor matched by each real frame, obtain the predicted frame matched by each real frame.
d. According to the coordinates of the real frame, the position of the real frame relative to the grid of the object is obtained, so as to calculate the error with the predicted coordinates.
